### PR TITLE
Fix race condition when sync with child and grandchild

### PR DESCRIPTION
**Pull request by hqhq**
Fixes: #1236

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

### CODE REVIEW

**Comment by hqhq** [comment]
ping @opencontainers/runc-maintainers

**Comment by teddyking** [comment]
fwiw it looks very likely that this PR also fixes #1281.

File: libcontainer/nsenter/nsexec.c [hunk 1]
```diff
@@ -695,6 +723,10 @@ void nsexec(void)
 				bail("failed to sync with parent: SYNC_RECVPID_ACK: got %u", s);
 			}
 
+			s = SYNC_CHILD_READY;
+			if (write(syncfd, &s, sizeof(s)) != sizeof(s))
+				bail("failed to sync with patent: write(SYNC_CHILD_READY)");
```
Comment (runcom) [inline libcontainer/nsenter/nsexec.c]: s/patent/parent/g

Comment (hqhq) [inline libcontainer/nsenter/nsexec.c]: Updated, thanks.

**Comment by crosbymichael** [comment]
@cyphar can you take a look at this one?

**Comment by stevenh** [comment]
I'm assuming this is in a final state now, if so then would be good to get it in ASAP as its causing pretty much all other PR's to have failed tests :(

**Comment by cyphar** [comment]
/me is looking at it. While it does fix the issue, there's some parts I'd like to get cleaned up before merging.

**Comment by stevenh** [comment]
Out of interest does that include changing the way the children are created and reaping the initial child which would make https://github.com/opencontainers/runc/pull/1301 unneeded?

**Comment by crosbymichael** [comment]
LGTM

**Comment by hqhq** [comment]
ping @cyphar

File: libcontainer/nsenter/nsexec.c [hunk 2]
```diff
@@ -570,20 +571,46 @@ void nsexec(void)
...
-					kill(child, SIGKILL);
-					bail("failed to sync with child: unexpected SYNC_RECVPID_ACK");
 					break;
 				case SYNC_CHILD_READY:
-					ready++;
+					ready = true;
 					break;
 				default:
-					bail("unexpected sync value");
+					bail("unexpected sync value: %u", s);
```
Comment (cyphar) [inline libcontainer/nsenter/nsexec.c]: While I understand this makes the code nicer, remember that the _only_ debugging information we normally get is the return code. Maybe we don't lose too much information, but something to keep in mind.

File: libcontainer/nsenter/nsexec.c [hunk 3]
```diff
@@ -570,20 +571,46 @@ void nsexec(void)
...
+				}
+			}
+
+			/* Now sync with grandchild. */
+
+			ready = false;
+			while (!ready) {
```
Comment (cyphar) [inline libcontainer/nsenter/nsexec.c]: What happens if the grandchild is ready before the child -- or if the child dies above but the grandchild is still around?

File: libcontainer/nsenter/nsexec.c [hunk 4]
```diff
@@ -413,7 +414,7 @@ void nsexec(void)
 {
 	int pipenum;
 	jmp_buf env;
-	int syncpipe[2];
+	int syncChildPipe[2], syncGrandchildPipe[2];
```
Comment (cyphar) [inline libcontainer/nsenter/nsexec.c]: Nit: Please use `snake_case` in C code.

File: libcontainer/nsenter/nsexec.c (hunk 3 above)
Comment (hqhq) [inline libcontainer/nsenter/nsexec.c]: > What happens if the grandchild is ready before the child

It won't, grandchild won't do anything before father finishes syncing with child and send SYNC_GRANDCHILD to grandchild.

> or if the child dies above but the grandchild is still around

All cases that child dies by error after creating grandchild, child will kill grandchild before exit.

File: libcontainer/nsenter/nsexec.c (hunk 2 above)
Comment (hqhq) [inline libcontainer/nsenter/nsexec.c]: Yeah, I don't think we lose any information here.

**Comment by hqhq** [comment]
Updated to use snake case.

@crosbymichael @cyphar PTAL.

**Comment by cyphar** [comment]
Failure is because `shfmt` is broken on Go 1.6.

LGTM

**Comment by cyphar** [comment]
I've made an issue (mvdan/sh#66) to track it.

**Comment by hqhq** [comment]
ping @crosbymichael

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**